packages/gatsby/src/utils/__tests__/mett.ts

```diff
@@ -0,0 +1,303 @@
+      met.off(`foo`, spy)
+      met.off(`foo`, spy)
+      met.on(`foo`, spy)
+      met.on(`foo`, spy)
+      met.on(`foo`, spy)
+      met.on(`foo`, spy)
+      met.on(`foo`, spy)
+      met.off(`foo`, spy)
+      met.off(`foo`, spy)
+      met.on(`foo`, spy)
+      met.on(`foo`, spy)
+      met.emit(`foo`, `pass`)
+
+      expect(spy).toHaveBeenCalledTimes(1)
+    })
+
+    it(`should allow same handler in different events`, (): void => {
+      const met = mett()
+      const spy = jest.fn()
+
+      met.on(`foo`, spy)
+      met.on(`bar`, spy)
+      met.emit(`foo`, `pass1`)
+      met.emit(`bar`, `pass2`)
+
+      expect(spy).toHaveBeenNthCalledWith(1, `pass1`, `foo`)
+      expect(spy).toHaveBeenNthCalledWith(2, `pass2`, `bar`)
+      expect(spy).toHaveBeenCalledTimes(2)
+    })
+  })
+
+  describe(`star events`, (): void => {
+    it(`should always call star handlers`, (): void => {
+      const met = mett()
+      const spy = jest.fn()
+
+      met.on(`*`, spy)
+      met.emit(`foo`, `pass1`)
+      met.emit(`bar`, `pass2`)
+
+      expect(spy).toHaveBeenNthCalledWith(1, `pass1`, `foo`)
+      expect(spy).toHaveBeenNthCalledWith(2, `pass2`, `bar`)
+      expect(spy).toHaveBeenCalledTimes(2)
+    })
+
+    it(`should call same handler twice if registered in star and event`, (): void => {
+      const met = mett()
+      const spy = jest.fn()
+
+      met.on(`bar`, spy)
+      met.on(`*`, spy)
+      met.emit(`foo`, `pass1`)
+      met.emit(`bar`, `pass2`)
+
+      expect(spy).toHaveBeenNthCalledWith(1, `pass1`, `foo`)
+      expect(spy).toHaveBeenNthCalledWith(2, `pass2`, `bar`)
+      expect(spy).toHaveBeenNthCalledWith(3, `pass2`, `bar`)
+      expect(spy).toHaveBeenCalledTimes(3)
+    })
+
+    it(`should not remove handler from star if removed from event`, (): void => {
+      const met = mett()
+      const spy = jest.fn()
+
+      met.on(`foo`, spy)
+      met.on(`*`, spy)
+      met.off(`foo`, spy)
+      met.emit(`foo`, `pass`)
+
+      expect(spy).toHaveBeenNthCalledWith(1, `pass`, `foo`)
+      expect(spy).toHaveBeenCalledTimes(1)
+    })
+
+    it(`should not remove handler from event if removed from star`, (): void => {
+      const met = mett()
+      const spy = jest.fn()
+
+      met.on(`*`, spy)
+      met.on(`foo`, spy)
+      met.off(`*`, spy)
+      met.emit(`foo`, `pass`)
+
+      expect(spy).toHaveBeenNthCalledWith(1, `pass`, `foo`)
+      expect(spy).toHaveBeenCalledTimes(1)
+    })
+
+    it(`should call star handlers after regular handlers`, (): void => {
+      const met = mett()
+
+      let out = ``
+
+      // Order is relevant here
+      met.on(`foo`, (): string => (out += `a`))
+      met.on(`*`, (): string => (out += `b`))
+      // Reverse order
+      met.on(`*`, (): string => (out += `c`))
+      met.on(`foo`, (): string => (out += `d`))
+
+      met.emit(`foo`, `pass`)
+
+      // the stars should last, in fifo order otherwise
+      expect(out).toBe(`adbc`)
+    })
+  })
+})
```

Comment:
Peter, these tests are amazing. Incredible coverage while also still being incredibly easy to parse. I feel like I fully understand this API just be reading these tests. 👍